File: benchmark.py
from __future__ import print_function

import platform
import sys
import time

from goless import backends, chan, go, selecting
from goless.compat import range


QUEUE_LEN = 10000
CHANSIZE_AND_NAMES = (
    (0, 'chan_sync'),
    (-1, 'chan_async'),
    (1000, 'chan_buff')
)


def bench_channel(chan_size):
    c = chan(chan_size)

    def func():
        for _ in range(QUEUE_LEN):
            c.send(0)
        c.close()
    count = 0

    go(func)
    start = time.clock()
    for _ in range(QUEUE_LEN):
        c.recv()
        count += 1
    end = time.clock()
    return end - start


def bench_channels():
    for size, name in CHANSIZE_AND_NAMES:
        took = bench_channel(size)
        write_result(name, took)


def bench_select(use_default):
    c = chan(0)
    cases = [
        selecting.scase(c, 1),
        selecting.rcase(c),
        selecting.scase(c, 1),
        selecting.rcase(c),
    ]
    if use_default:
        cases.append(selecting.dcase())

    def sender():
        while True:
            c.send(0)
            c.recv()
    go(sender)

    start = time.clock()
    for _ in range(QUEUE_LEN):
        selecting.select(cases)
    end = time.clock()
    return end - start


def bench_selects():
    took_nodefault = bench_select(False)
    write_result('select', took_nodefault)
    took_withdefault = bench_select(True)
    write_result('select_default', took_withdefault)


WRITE_ENABLED = True
PYIMPL = '%s%s' % (platform.python_implementation(), sys.version_info[0])


def write_result(benchname, elapsed):
    if not WRITE_ENABLED:
        return
    w = sys.stdout.write
    w(PYIMPL)
    w(' ')
    w(backends.current.shortname())
    w(' ')
    w(benchname)
    w(' ')
    w(('%.5f' % elapsed))
    w('\n')


def prime():
    count = 1
    if platform.python_implementation() == 'PyPy':
        count = 10
    global WRITE_ENABLED
    WRITE_ENABLED = False
    for _ in range(count):
        bench_channels()
        bench_selects()
    WRITE_ENABLED = True


def main():
    prime()
    bench_channels()
    bench_selects()


if __name__ == '__main__':
    main()

File: benchmark.go
/*
	Golang implementation of this project's benchmark.py file.
	It should be a pretty direct mirror.

	The only questionable decision is the use of reflect.SelectCase,
	rather than Go's select statement.
	goless' select function is based on reflect.SelectCase,
	so this is "idiotmatic goless" (ugh),
	but Go programs rarely use reflect.SelectCase and instead use its
	select function.

	Using the select statement would probably speed things up.
	I'm not sure what the right thing to do is,
	so I chose the easier thing.
*/

package main

import (
	"fmt"
	"reflect"
	"runtime"
	"time"
)

const queueLen int = 10000
type timing float64

func benchChannel(chanSize int) timing {
	c := make(chan int, chanSize)
	go func() {
		for i := 0; i <= queueLen; i++ {
			c <- 0
		}
		close(c)
	}()

	start := time.Now()
	for i := 0; i < queueLen; i++ {
		<-c
	}
	elapsed := time.Since(start)
	return timing(elapsed.Seconds())
}

func benchChannels() {
	tookSync := benchChannel(0)
	writeResult("chan_sync", tookSync)
	tookAsync := benchChannel(queueLen + 1)
	writeResult("chan_async", tookAsync)
	tookBuff := benchChannel(1000)
	writeResult("chan_buff", tookBuff)
}

func benchSelect(useDefault bool) timing {
    c := make(chan int, 0)

	caseCnt := 4
	if useDefault {
		caseCnt = 5
	}
	cases := make([]reflect.SelectCase, caseCnt)
	cases[0] = reflect.SelectCase{Dir: reflect.SelectSend, Chan: reflect.ValueOf(c), Send: reflect.ValueOf(0)}
	cases[1] = reflect.SelectCase{Dir: reflect.SelectRecv, Chan: reflect.ValueOf(c)}
	cases[2] = cases[0]
	cases[3] = cases[1]
	if useDefault {
		cases[4] = reflect.SelectCase{Dir: reflect.SelectDefault}
	}

    go func() {
		for {
			c <- 0
			<- c
		}
	}()

	start := time.Now()
	for i := 0; i <= queueLen; i++ {
		reflect.Select(cases)
	}
	elapsed := time.Since(start)
	return timing(elapsed.Seconds())
}

func benchSelects() {
	tookNoDefault := benchSelect(false)
	writeResult("select", tookNoDefault)
	tookWithDefault := benchSelect(true)
	writeResult("select_default", tookWithDefault)
}

func writeResult(benchName string, elapsed timing) {
	fmt.Printf("go %s %s %.5f\n", runtime.Compiler, benchName, elapsed)
}

func main() {
	benchChannels()
	benchSelects()
}
